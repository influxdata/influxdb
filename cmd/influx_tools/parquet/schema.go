package parquet

import (
	"context"
	"fmt"
	"sort"
	"strings"

	"github.com/apache/arrow/go/v16/arrow"

	"github.com/influxdata/influxdb/models"
	"github.com/influxdata/influxdb/tsdb"
	"github.com/influxdata/influxql"
)

type seriesEntry struct {
	key    string
	tags   models.Tags
	fields map[string]influxql.DataType
}

type schemaCreator struct {
	measurement string
	shards      []*tsdb.Shard

	series map[uint64][]seriesEntry

	tags      []string
	fields    map[string]influxql.DataType
	fieldKeys []string
	conflicts map[string][]influxql.DataType

	typeResolutions map[string]influxql.DataType
	nameResolutions map[string]string
}

func (s *schemaCreator) extractSchema(ctx context.Context) error {
	// Iterate over the shards and extract all series
	for _, shard := range s.shards {
		// Extract all fields of the measurement
		fields := shard.MeasurementFields([]byte(s.measurement)).FieldSet()
		if len(fields) == 0 {
			continue
		}

		// Collect all available series in the shard and measurement and store
		// them for later use and series accumulation
		seriesCursor, err := shard.CreateSeriesCursor(
			ctx,
			tsdb.SeriesCursorRequest{},
			influxql.MustParseExpr("_name = '"+s.measurement+"'"),
		)
		if err != nil {
			return fmt.Errorf("getting series cursor failed: %w", err)
		}
		defer seriesCursor.Close()

		for {
			cur, err := seriesCursor.Next()
			if err != nil {
				return fmt.Errorf("advancing series cursor failed: %w", err)
			}
			if cur == nil {
				break
			}
			mname := string(cur.Name)
			if mname != s.measurement {
				continue
			}

			s.series[shard.ID()] = append(s.series[shard.ID()], seriesEntry{
				key:    s.measurement + "." + string(cur.Tags.HashKey(true)),
				tags:   cur.Tags.Clone(),
				fields: fields,
			})
		}
	}

	// Collect all tags and fields for creating the overall schema
	tags := make(map[string]bool)
	conflicts := make(map[string]map[influxql.DataType]bool)
	s.fields = make(map[string]influxql.DataType)
	for _, series := range s.series {
		for _, serie := range series {
			// Dedup tag keys
			for _, tag := range serie.tags {
				tags[string(tag.Key)] = true
			}
			// Detect field type conflicts and collect the fields
			for name, current := range serie.fields {
				if existing, found := s.fields[name]; found && current != existing {
					if _, exists := conflicts[name]; !exists {
						conflicts[name] = make(map[influxql.DataType]bool)
					}
					conflicts[name][current] = true
					conflicts[name][existing] = true
				}
				s.fields[name] = current
			}
		}
	}

	// Apply the type resolutions
	for name, ftype := range s.typeResolutions {
		if _, found := s.fields[name]; !found {
			continue
		}
		s.fields[name] = ftype
	}

	// Boil down all tags, fields and conflicts for later reuse
	s.tags = make([]string, 0, len(tags))
	for name := range tags {
		s.tags = append(s.tags, name)
	}
	sort.Strings(s.tags)

	s.fieldKeys = make([]string, 0, len(s.fields))
	for name, ftype := range s.fields {
		// Detect unconvertible fields
		switch ftype {
		case influxql.Float, influxql.Integer, influxql.String, influxql.Boolean, influxql.Unsigned:
			// Accepted type
			s.fieldKeys = append(s.fieldKeys, name)
		default:
			// Unconvertible type
			panic(fmt.Errorf("unconvertible field %q with type %v", name, ftype))
		}
	}
	sort.Strings(s.fieldKeys)

	s.conflicts = make(map[string][]influxql.DataType, len(conflicts))
	for name, conflict := range conflicts {
		s.conflicts[name] = make([]influxql.DataType, 0, len(conflict))
		for ftype := range conflict {
			s.conflicts[name] = append(s.conflicts[name], ftype)
		}
	}

	return nil
}

func (s *schemaCreator) validate() (bool, error) {
	var hasConflicts bool
	var typeConflicts []string
	for field := range s.conflicts {
		hasConflicts = true
		if _, resolved := s.typeResolutions[field]; !resolved {
			typeConflicts = append(typeConflicts, field)
		}
	}
	if len(typeConflicts) > 0 {
		return true, fmt.Errorf("unresolved type conflicts for %q", strings.Join(typeConflicts, ","))
	}

	var nameConflicts []string
	for _, field := range s.fieldKeys {
		for _, tag := range s.tags {
			if tag == field {
				hasConflicts = true
				if _, resolved := s.nameResolutions[field]; !resolved {
					nameConflicts = append(nameConflicts, tag)
				}
			}
		}
	}
	if len(nameConflicts) > 0 {
		return true, fmt.Errorf("unresolved name conflicts for %q", strings.Join(nameConflicts, ","))
	}

	return hasConflicts, nil
}

func (s *schemaCreator) schema(info map[string]string) (*arrow.Schema, error) {
	columns := make([]arrow.Field, 0, 1+len(s.tags)+len(s.fields))

	// Add the timestamp column first
	columns = append(columns, arrow.Field{
		Name:     "time",
		Type:     &arrow.TimestampType{Unit: arrow.Nanosecond},
		Metadata: arrow.MetadataFrom(map[string]string{"kind": "timestamp"}),
	})

	// Add tags in alphabetical order
	for _, tag := range s.tags {
		columns = append(columns, arrow.Field{
			Name:     tag,
			Type:     &arrow.StringType{},
			Nullable: true,
			Metadata: arrow.MetadataFrom(map[string]string{"kind": "tag"}),
		})
	}

	// Add fields in alphabetical order with the corresponding arrow type
	for _, name := range s.fieldKeys {
		ftype := s.fields[name]
		if t, found := s.typeResolutions[name]; found {
			ftype = t
		}
		fname := name
		if n, found := s.nameResolutions[name]; found {
			fname = n
		}

		var dtype arrow.DataType
		switch ftype {
		case influxql.Float:
			dtype = &arrow.Float64Type{}
		case influxql.Integer:
			dtype = &arrow.Int64Type{}
		case influxql.String:
			dtype = &arrow.StringType{}
		case influxql.Boolean:
			dtype = &arrow.BooleanType{}
		case influxql.Unsigned:
			dtype = &arrow.Uint64Type{}
		default:
			panic(fmt.Errorf("unconvertible field %q", name))
		}
		columns = append(columns, arrow.Field{
			Name:     fname,
			Type:     dtype,
			Nullable: true,
			Metadata: arrow.MetadataFrom(map[string]string{"kind": "field"}),
		})
	}

	// Add the metadata given as argument and add info about column kinds
	if info == nil {
		info = make(map[string]string)
	}
	info["tags"] = strings.Join(s.tags, ",")
	info["fields"] = strings.Join(s.fieldKeys, ",")
	metadata := arrow.MetadataFrom(info)

	return arrow.NewSchema(columns, &metadata), nil
}
