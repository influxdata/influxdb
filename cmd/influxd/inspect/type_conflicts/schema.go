package typecheck

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"

	errors2 "github.com/influxdata/influxdb/v2/pkg/errors"
)

type UniqueField struct {
	Database    string `json:"database"`
	Retention   string `json:"retention"`
	Measurement string `json:"measurement"`
	Field       string `json:"field"`
}

type FieldTypes map[string]struct{}
type Schema map[string]FieldTypes

func (ft FieldTypes) MarshalText() (text []byte, err error) {
	s := make([]string, 0, len(ft))
	for f := range ft {
		s = append(s, f)
	}
	return []byte(strings.Join(s, ",")), nil
}

func (ft *FieldTypes) UnmarshalText(text []byte) error {
	if *ft == nil {
		*ft = make(FieldTypes)
	}
	for _, ty := range strings.Split(string(text), ",") {
		(*ft)[ty] = struct{}{}
	}
	return nil
}

func NewSchema() Schema {
	return make(Schema)
}

func SchemaFromFile(filename string) (Schema, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("unable to open schema file %q: %w", filename, err)
	}

	s := NewSchema()
	if err := s.Decode(f); err != nil {
		return nil, fmt.Errorf("unable to decode schema file %q: %w", filename, err)
	}
	return s, nil
}

func (uf *UniqueField) String() string {
	return fmt.Sprintf("%q.%q.%q.%q", uf.Database, uf.Retention, uf.Measurement, uf.Field)
}

func (s Schema) AddField(database, retention, measurement, field, dataType string) {
	uf := UniqueField{
		Database:    database,
		Retention:   retention,
		Measurement: measurement,
		Field:       field,
	}
	s.AddFormattedField(uf.String(), dataType)
}

func (s Schema) AddFormattedField(field string, dataType string) {
	if _, ok := s[field]; !ok {
		s[field] = make(map[string]struct{})
	}
	s[field][dataType] = struct{}{}
}

func (s Schema) Merge(schema Schema) {
	for field, types := range schema {
		for t := range types {
			s.AddFormattedField(field, t)
		}
	}
}

func (s Schema) Conflicts() Schema {
	cs := NewSchema()
	for field, t := range s {
		if len(t) > 1 {
			for ty := range t {
				cs.AddFormattedField(field, ty)
			}
		}
	}
	return cs
}

func (s Schema) WriteSchemaFile(filename string) error {
	if len(s) == 0 {
		fmt.Println("No schema file generated: no valid measurements/fields found")
		return nil
	}

	if err := s.encodeSchema(filename); err != nil {
		return fmt.Errorf("unable to write schema file to %q: %w", filename, err)
	}
	fmt.Printf("Schema file written successfully to: %q\n", filename)
	return nil
}

func (s Schema) WriteConflictsFile(filename string) error {
	conflicts := s.Conflicts()
	if len(conflicts) == 0 {
		fmt.Println("No conflicts file generated: no conflicts found")
		return nil
	}

	if err := conflicts.encodeSchema(filename); err != nil {
		return fmt.Errorf("unable to write conflicts file to %q: %w", filename, err)
	}
	fmt.Printf("Conflicts file written successfully to: %q\n", filename)
	return nil
}

func (s Schema) encodeSchema(filename string) (rErr error) {
	schemaFile, err := os.Create(filename)
	defer errors2.Capture(&rErr, schemaFile.Close)()
	if err != nil {
		return fmt.Errorf("unable to create schema file: %w", err)
	}
	return s.Encode(schemaFile)
}

func (s Schema) Encode(w io.Writer) error {
	enc := json.NewEncoder(w)
	enc.SetIndent("", "  ")
	if err := enc.Encode(s); err != nil {
		return fmt.Errorf("unable to encode schema: %w", err)
	}
	return nil
}

func (s Schema) Decode(r io.Reader) error {
	if err := json.NewDecoder(r).Decode(&s); err != nil {
		return fmt.Errorf("unable to decode schema: %w", err)
	}
	return nil
}
