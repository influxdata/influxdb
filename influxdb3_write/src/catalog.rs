//! Implementation of the Catalog that sits entirely in memory.

use crate::SequenceNumber;
use influxdb_line_protocol::FieldValue;
use observability_deps::tracing::info;
use parking_lot::RwLock;
use schema::{InfluxColumnType, InfluxFieldType, Schema, SchemaBuilder};
use serde::{Deserialize, Serialize, Serializer};
use std::collections::{BTreeMap, HashMap};
use std::sync::Arc;
use thiserror::Error;

mod serialize;

#[derive(Debug, Error)]
pub enum Error {
    #[error("catalog updated elsewhere")]
    CatalogUpdatedElsewhere,

    #[error(
        "Update to schema would exceed number of columns per table limit of {} columns",
        Catalog::NUM_COLUMNS_PER_TABLE_LIMIT - 1
    )]
    TooManyColumns,

    #[error(
        "Update to schema would exceed number of tables limit of {} tables",
        Catalog::NUM_TABLES_LIMIT
    )]
    TooManyTables,

    #[error(
        "Adding a new database would exceed limit of {} databases",
        Catalog::NUM_DBS_LIMIT
    )]
    TooManyDbs,

    #[error("last cache size must be from 1 to 10")]
    InvalidLastCacheSize,
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

pub const TIME_COLUMN_NAME: &str = "time";

#[derive(Debug)]
pub struct Catalog {
    inner: RwLock<InnerCatalog>,
}

impl Default for Catalog {
    fn default() -> Self {
        Self::new()
    }
}

impl PartialEq for Catalog {
    fn eq(&self, other: &Self) -> bool {
        self.inner.read().eq(&other.inner.read())
    }
}

impl Serialize for Catalog {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.inner.read().serialize(serializer)
    }
}

impl Catalog {
    /// Limit for the number of Databases that InfluxDB Edge can have
    pub(crate) const NUM_DBS_LIMIT: usize = 5;
    /// Limit for the number of columns per table that InfluxDB Edge can have
    pub(crate) const NUM_COLUMNS_PER_TABLE_LIMIT: usize = 500;
    /// Limit for the number of tables across all DBs that InfluxDB Edge can have
    pub(crate) const NUM_TABLES_LIMIT: usize = 2000;

    pub fn new() -> Self {
        Self {
            inner: RwLock::new(InnerCatalog::new()),
        }
    }

    pub fn from_inner(inner: InnerCatalog) -> Self {
        Self {
            inner: RwLock::new(inner),
        }
    }

    pub(crate) fn replace_database(
        &self,
        sequence: SequenceNumber,
        db: Arc<DatabaseSchema>,
    ) -> Result<()> {
        let mut inner = self.inner.write();
        if inner.sequence != sequence {
            info!("catalog updated elsewhere");
            return Err(Error::CatalogUpdatedElsewhere);
        }

        // Check we have not gone over the table limit with this updated DB
        let mut num_tables = inner
            .databases
            .iter()
            .filter(|(k, _)| *k != &db.name)
            .map(|(_, v)| v)
            .fold(0, |acc, db| acc + db.tables.len());

        num_tables += db.tables.len();

        if num_tables > Self::NUM_TABLES_LIMIT {
            return Err(Error::TooManyTables);
        }

        for table in db.tables.values() {
            if table.num_columns() > Self::NUM_COLUMNS_PER_TABLE_LIMIT {
                return Err(Error::TooManyColumns);
            }
        }

        info!("inserted/updated database in catalog: {}", db.name);
        inner.sequence = inner.sequence.next();
        inner.databases.insert(db.name.clone(), db);
        Ok(())
    }

    pub(crate) fn db_or_create(
        &self,
        db_name: &str,
    ) -> Result<(SequenceNumber, Arc<DatabaseSchema>)> {
        let (sequence, db) = {
            let inner = self.inner.read();
            (inner.sequence, inner.databases.get(db_name).cloned())
        };

        let db = match db {
            Some(db) => {
                info!("return existing db {}", db_name);
                db
            }
            None => {
                let mut inner = self.inner.write();

                if inner.databases.len() >= Self::NUM_DBS_LIMIT {
                    return Err(Error::TooManyDbs);
                }

                info!("return new db {}", db_name);
                let db = Arc::new(DatabaseSchema::new(db_name));
                inner.databases.insert(db.name.clone(), Arc::clone(&db));
                db
            }
        };

        Ok((sequence, db))
    }

    pub fn db_schema(&self, name: &str) -> Option<Arc<DatabaseSchema>> {
        info!("db_schema {}", name);
        self.inner.read().databases.get(name).cloned()
    }

    pub fn sequence_number(&self) -> SequenceNumber {
        self.inner.read().sequence
    }

    pub fn clone_inner(&self) -> InnerCatalog {
        self.inner.read().clone()
    }

    pub fn list_databases(&self) -> Vec<String> {
        self.inner.read().databases.keys().cloned().collect()
    }

    #[cfg(test)]
    pub fn db_exists(&self, db_name: &str) -> bool {
        self.inner.read().db_exists(db_name)
    }
}

#[serde_with::serde_as]
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone, Default)]
pub struct InnerCatalog {
    /// The catalog is a map of databases with their table schemas
    #[serde_as(as = "serde_with::MapPreventDuplicates<_, _>")]
    databases: HashMap<String, Arc<DatabaseSchema>>,
    sequence: SequenceNumber,
}

impl InnerCatalog {
    pub(crate) fn new() -> Self {
        Self {
            databases: HashMap::new(),
            sequence: SequenceNumber::new(0),
        }
    }

    pub fn sequence_number(&self) -> SequenceNumber {
        self.sequence
    }

    #[cfg(test)]
    pub fn db_exists(&self, db_name: &str) -> bool {
        self.databases.contains_key(db_name)
    }
}

#[serde_with::serde_as]
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone)]
pub struct DatabaseSchema {
    pub name: String,
    /// The database is a map of tables
    #[serde_as(as = "serde_with::MapPreventDuplicates<_, _>")]
    pub(crate) tables: BTreeMap<String, TableDefinition>,
}

impl DatabaseSchema {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            tables: BTreeMap::new(),
        }
    }

    pub fn get_table_schema(&self, table_name: &str) -> Option<&Schema> {
        self.tables.get(table_name).map(|table| &table.schema)
    }

    pub fn get_table(&self, table_name: &str) -> Option<&TableDefinition> {
        self.tables.get(table_name)
    }

    pub fn table_names(&self) -> Vec<String> {
        self.tables.keys().cloned().collect()
    }

    pub fn table_exists(&self, table_name: &str) -> bool {
        self.tables.contains_key(table_name)
    }
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct TableDefinition {
    pub name: String,
    pub schema: Schema,
    pub last_caches: Vec<LastCacheDefinition>,
}

impl TableDefinition {
    /// Create a new [`TableDefinition`]
    ///
    /// Ensures the the provided columns will be ordered before constructing the schema.
    pub(crate) fn new<N: Into<String>, CN: AsRef<str>>(
        name: N,
        columns: impl AsRef<[(CN, InfluxColumnType)]>,
        series_key: Option<impl IntoIterator<Item: AsRef<str>>>,
    ) -> Self {
        // Use a BTree to ensure that the columns are ordered:
        let mut ordered_columns = BTreeMap::new();
        for (name, column_type) in columns.as_ref() {
            ordered_columns.insert(name.as_ref(), column_type);
        }
        let mut schema_builder = SchemaBuilder::with_capacity(columns.as_ref().len());
        let name = name.into();
        schema_builder.measurement(&name);
        if let Some(sk) = series_key {
            schema_builder.with_series_key(sk);
        }
        for (name, column_type) in ordered_columns {
            schema_builder.influx_column(name, *column_type);
        }
        let schema = schema_builder.build().unwrap();

        Self {
            name,
            schema,
            last_caches: vec![],
        }
    }

    /// Check if the column exists in the [`TableDefinition`]s schema
    pub(crate) fn column_exists(&self, column: &str) -> bool {
        self.schema.find_index_of(column).is_some()
    }

    /// Add the columns to this [`TableDefinition`]
    ///
    /// This ensures that the resulting schema has its columns ordered
    pub(crate) fn add_columns(&mut self, columns: Vec<(String, InfluxColumnType)>) {
        // Use BTree to insert existing and new columns, and use that to generate the
        // resulting schema, to ensure column order is consistent:
        let mut cols = BTreeMap::new();
        for (col_type, field) in self.schema.iter() {
            cols.insert(field.name(), col_type);
        }
        for (name, column_type) in columns.iter() {
            cols.insert(name, *column_type);
        }
        let mut schema_builder = SchemaBuilder::with_capacity(columns.len());
        // TODO: may need to capture some schema-level metadata, currently, this causes trouble in
        // tests, so I am omitting this for now:
        // schema_builder.measurement(&self.name);
        for (name, col_type) in cols {
            schema_builder.influx_column(name, col_type);
        }
        let schema = schema_builder.build().unwrap();

        self.schema = schema;
    }

    pub(crate) fn index_columns(&self) -> Vec<&str> {
        self.schema
            .iter()
            .filter_map(|(col_type, field)| match col_type {
                InfluxColumnType::Tag => Some(field.name().as_str()),
                InfluxColumnType::Field(_) | InfluxColumnType::Timestamp => None,
            })
            .collect()
    }

    pub(crate) fn schema(&self) -> &Schema {
        &self.schema
    }

    pub(crate) fn num_columns(&self) -> usize {
        self.schema.len()
    }

    pub(crate) fn field_type_by_name(&self, name: &str) -> Option<InfluxColumnType> {
        self.schema.field_type_by_name(name)
    }

    pub(crate) fn is_v3(&self) -> bool {
        self.schema.series_key().is_some()
    }

    /// Add a new last cache to this table definition
    #[cfg(test)]
    pub(crate) fn add_last_cache<L>(&mut self, last_cache: L)
    where
        L: Into<LastCacheDefinition>,
    {
        self.last_caches.push(last_cache.into());
    }
}

/// Defines a last cache in a given table and database
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone)]
pub struct LastCacheDefinition {
    /// Given name of the cache
    pub name: String,
    /// Columns intended to be used as predicates in the cache
    pub key_columns: Vec<String>,
    /// Columns that store values in the cache
    pub value_columns: Vec<String>,
    /// The number of last values to hold in the cache
    count: LastCacheSize,
}

impl LastCacheDefinition {
    /// Create a new [`LastCacheDefinition`]
    #[cfg(test)]
    pub(crate) fn new<N, K, V>(
        name: N,
        key_columns: K,
        value_columns: V,
        count: usize,
    ) -> Result<Self, Error>
    where
        N: Into<String>,
        K: IntoIterator<Item: Into<String>>,
        V: IntoIterator<Item: Into<String>>,
    {
        Ok(Self {
            name: name.into(),
            key_columns: key_columns.into_iter().map(Into::into).collect(),
            value_columns: value_columns.into_iter().map(Into::into).collect(),
            count: count.try_into()?,
        })
    }
}

/// The maximum allowed size for a last cache
const LAST_CACHE_MAX_SIZE: usize = 10;

/// The size of the last cache
///
/// Must be between 1 and [`LAST_CACHE_MAX_SIZE`]
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone, Copy)]
pub struct LastCacheSize(usize);

impl LastCacheSize {
    pub fn new(size: usize) -> Result<Self, Error> {
        if size == 0 || size > LAST_CACHE_MAX_SIZE {
            Err(Error::InvalidLastCacheSize)
        } else {
            Ok(Self(size))
        }
    }
}

impl TryFrom<usize> for LastCacheSize {
    type Error = Error;

    fn try_from(value: usize) -> Result<Self, Self::Error> {
        Self::new(value)
    }
}

impl From<LastCacheSize> for usize {
    fn from(value: LastCacheSize) -> Self {
        value.0
    }
}

pub fn influx_column_type_from_field_value(fv: &FieldValue<'_>) -> InfluxColumnType {
    match fv {
        FieldValue::I64(_) => InfluxColumnType::Field(InfluxFieldType::Integer),
        FieldValue::U64(_) => InfluxColumnType::Field(InfluxFieldType::UInteger),
        FieldValue::F64(_) => InfluxColumnType::Field(InfluxFieldType::Float),
        FieldValue::String(_) => InfluxColumnType::Field(InfluxFieldType::String),
        FieldValue::Boolean(_) => InfluxColumnType::Field(InfluxFieldType::Boolean),
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use pretty_assertions::assert_eq;
    use test_helpers::assert_contains;

    use super::*;

    type SeriesKey = Option<Vec<String>>;

    #[test]
    fn catalog_serialization() {
        let catalog = Catalog::new();
        let mut database = DatabaseSchema {
            name: "test_db".to_string(),
            tables: BTreeMap::new(),
        };
        use InfluxColumnType::*;
        use InfluxFieldType::*;
        database.tables.insert(
            "test_table_1".into(),
            TableDefinition::new(
                "test_table_1",
                [
                    ("tag_1", Tag),
                    ("tag_2", Tag),
                    ("tag_3", Tag),
                    ("time", Timestamp),
                    ("string_field", Field(String)),
                    ("bool_field", Field(Boolean)),
                    ("i64_field", Field(Integer)),
                    ("u64_field", Field(UInteger)),
                    ("f64_field", Field(Float)),
                ],
                SeriesKey::None,
            ),
        );
        database.tables.insert(
            "test_table_2".into(),
            TableDefinition::new(
                "test_table_2",
                [
                    ("tag_1", Tag),
                    ("tag_2", Tag),
                    ("tag_3", Tag),
                    ("time", Timestamp),
                    ("string_field", Field(String)),
                    ("bool_field", Field(Boolean)),
                    ("i64_field", Field(Integer)),
                    ("u64_field", Field(UInteger)),
                    ("f64_field", Field(Float)),
                ],
                SeriesKey::None,
            ),
        );
        let database = Arc::new(database);
        catalog
            .replace_database(SequenceNumber::new(0), database)
            .unwrap();

        // Perform a snapshot test to check that the JSON serialized catalog does not change in an
        // undesired way when introducing features etc.
        assert_json_snapshot!(catalog);

        // Serialize/deserialize to ensure roundtrip to/from JSON
        let serialized = serde_json::to_string(&catalog).unwrap();
        let deserialized_inner: InnerCatalog = serde_json::from_str(&serialized).unwrap();
        let deserialized = Catalog::from_inner(deserialized_inner);
        assert_eq!(catalog, deserialized);
    }

    #[test]
    fn invalid_catalog_deserialization() {
        // Duplicate databases
        {
            let json = r#"{
                "databases": {
                    "db1": {
                        "name": "db1",
                        "tables": {}
                    },
                    "db1": {
                        "name": "db1",
                        "tables": {}
                    }
                }
            }"#;
            let err = serde_json::from_str::<InnerCatalog>(json).unwrap_err();
            assert_contains!(err.to_string(), "found duplicate key");
        }
        // Duplicate tables
        {
            let json = r#"{
                "databases": {
                    "db1": {
                        "name": "db1",
                        "tables": {
                            "tbl1": {
                                "name": "tbl1",
                                "cols": {}
                            },
                            "tbl1": {
                                "name": "tbl1",
                                "cols": {}
                            }
                        }
                    }
                }
            }"#;
            let err = serde_json::from_str::<InnerCatalog>(json).unwrap_err();
            assert_contains!(err.to_string(), "found duplicate key");
        }
        // Duplicate columns
        {
            let json = r#"{
                "databases": {
                    "db1": {
                        "name": "db1",
                        "tables": {
                            "tbl1": {
                                "name": "tbl1",
                                "cols": {
                                    "col1": {
                                        "type": "i64",
                                        "influx_type": "field",
                                        "nullable": true
                                    },
                                    "col1": {
                                        "type": "u64",
                                        "influx_type": "field",
                                        "nullable": true
                                    }
                                }
                            }
                        }
                    }
                }
            }"#;
            let err = serde_json::from_str::<InnerCatalog>(json).unwrap_err();
            assert_contains!(err.to_string(), "found duplicate key");
        }
    }

    #[test]
    fn add_columns_updates_schema() {
        let mut database = DatabaseSchema {
            name: "test".to_string(),
            tables: BTreeMap::new(),
        };
        database.tables.insert(
            "test".into(),
            TableDefinition::new(
                "test",
                [(
                    "test".to_string(),
                    InfluxColumnType::Field(InfluxFieldType::String),
                )],
                SeriesKey::None,
            ),
        );

        let table = database.tables.get_mut("test").unwrap();
        table.add_columns(vec![("test2".to_string(), InfluxColumnType::Tag)]);
        let schema = table.schema();
        assert_eq!(
            schema.field(0).0,
            InfluxColumnType::Field(InfluxFieldType::String)
        );
        assert_eq!(schema.field(1).0, InfluxColumnType::Tag);
    }

    #[test]
    fn serialize_series_keys() {
        let catalog = Catalog::new();
        let mut database = DatabaseSchema {
            name: "test_db".to_string(),
            tables: BTreeMap::new(),
        };
        use InfluxColumnType::*;
        use InfluxFieldType::*;
        database.tables.insert(
            "test_table_1".into(),
            TableDefinition::new(
                "test_table_1",
                [
                    ("tag_1", Tag),
                    ("tag_2", Tag),
                    ("tag_3", Tag),
                    ("time", Timestamp),
                    ("field", Field(String)),
                ],
                SeriesKey::Some(vec![
                    "tag_1".to_string(),
                    "tag_2".to_string(),
                    "tag_3".to_string(),
                ]),
            ),
        );
        let database = Arc::new(database);
        catalog
            .replace_database(SequenceNumber::new(0), database)
            .unwrap();

        assert_json_snapshot!(catalog);

        let serialized = serde_json::to_string(&catalog).unwrap();
        let deserialized_inner: InnerCatalog = serde_json::from_str(&serialized).unwrap();
        let deserialized = Catalog::from_inner(deserialized_inner);
        assert_eq!(catalog, deserialized);
    }

    #[test]
    fn serialize_last_cache() {
        let catalog = Catalog::new();
        let mut database = DatabaseSchema {
            name: "test_db".to_string(),
            tables: BTreeMap::new(),
        };
        use InfluxColumnType::*;
        use InfluxFieldType::*;
        let mut table_def = TableDefinition::new(
            "test",
            [
                ("tag_1", Tag),
                ("tag_2", Tag),
                ("tag_3", Tag),
                ("time", Timestamp),
                ("field", Field(String)),
            ],
            SeriesKey::None,
        );
        table_def.add_last_cache(
            LastCacheDefinition::new("test_table_last_cache", ["tag_2", "tag_3"], ["field"], 1)
                .unwrap(),
        );
        database.tables.insert("test_table_1".into(), table_def);
        let database = Arc::new(database);
        catalog
            .replace_database(SequenceNumber::new(0), database)
            .unwrap();

        assert_json_snapshot!(catalog);

        let serialized = serde_json::to_string(&catalog).unwrap();
        let deserialized_inner: InnerCatalog = serde_json::from_str(&serialized).unwrap();
        let deserialized = Catalog::from_inner(deserialized_inner);
        assert_eq!(catalog, deserialized);
    }
}
