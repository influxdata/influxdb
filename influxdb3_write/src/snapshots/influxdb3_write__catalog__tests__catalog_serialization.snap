---
source: influxdb3_write/src/catalog.rs
expression: catalog
---
{
  "databases": {
    "test_db": {
      "name": "test_db",
      "tables": {
        "test_table_1": {
          "name": "test_table_1",
          "cols": {
            "bool_field": {
              "type": "bool",
              "influx_type": "field",
              "nullable": true
            },
            "f64_field": {
              "type": "f64",
              "influx_type": "field",
              "nullable": true
            },
            "i64_field": {
              "type": "i64",
              "influx_type": "field",
              "nullable": true
            },
            "string_field": {
              "type": "str",
              "influx_type": "field",
              "nullable": true
            },
            "tag_1": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "tag_2": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "tag_3": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "time": {
              "type": {
                "time": [
                  "ns",
                  null
                ]
              },
              "influx_type": "time",
              "nullable": false
            },
            "u64_field": {
              "type": "u64",
              "influx_type": "field",
              "nullable": true
            }
          }
        },
        "test_table_2": {
          "name": "test_table_2",
          "cols": {
            "bool_field": {
              "type": "bool",
              "influx_type": "field",
              "nullable": true
            },
            "f64_field": {
              "type": "f64",
              "influx_type": "field",
              "nullable": true
            },
            "i64_field": {
              "type": "i64",
              "influx_type": "field",
              "nullable": true
            },
            "string_field": {
              "type": "str",
              "influx_type": "field",
              "nullable": true
            },
            "tag_1": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "tag_2": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "tag_3": {
              "type": {
                "dict": [
                  "i32",
                  "str"
                ]
              },
              "influx_type": "tag",
              "nullable": true
            },
            "time": {
              "type": {
                "time": [
                  "ns",
                  null
                ]
              },
              "influx_type": "time",
              "nullable": false
            },
            "u64_field": {
              "type": "u64",
              "influx_type": "field",
              "nullable": true
            }
          }
        }
      }
    }
  },
  "sequence": 0
}
